### DISCUSSION TITLE
The library is not replaced after reinstalling the module

### DISCUSSION THREAD

**Discussion by VitalyaK** [dnnsoftware/Dnn.Platform#4616] [General] [answered]
I find that the Extension Installation is not working at all : the DLL file in *_install.zip is not updated during Replace Module

**Steps to reproduce:**
- login with admin account
- Go to 'Settings/Extensions' (by side bar)
- Click on 'Install Extension'
- Select an existed extension
- continue the 5 steps

**Current behavior**
> . . . 
> Info Starting Installation - Assembly
> Info Assembly already registered - bin\MyLibrary.dll
> Info Component installed successfully - Assembly
> Info Installation committed
> . . . 

**Expected behavior**
> . . . 
> Info Starting Installation - Assembly
> Info Assembly already registered - bin\MyLibrary.dll
> Info Creating backup of previous version - bin\MyLibrary.dll
> Info Created - bin\MyLibrary.dll
> Info Component installed successfully - Assembly
> Info Installation committed
> . . . 

I don't understand why this is happening ((
Previously, this always replaced all libraries.
Or maybe I can somehow force the system to always replace the library when installing the module ?

I also noticed that in the .dnn manifest file, I can specify the version for the library.
If the version of the library is not specified, then after reinstalling the module, the library is not replaced ((
But if I specify the version of the library,
>             <assembly>
>               <path>bin</path>
>               <name>MyLibrary.dll</name>
>               <sourceFileName>bin/MyLibrary.dll</sourceFileName>
>               <version>99.99.99.99</version>
>             </assembly>
then it will always replace the library

I don't understand, can I specify a large version so that the libraries are always replaced ?

**Comment by valadas** [comment, marked as answer]
It is best practice to always declare the actual version. Now I believe if the version is the same or higher it will install it but if the version is lower it will skip it. I am not 100% sure though, can you try just setting the same version on the first and second run and see if that works ?

**Reply by VitalyaK** [reply]
If the library does not have a version, then it skips. Why ? 
It seems to me that reinstallation should replace all dll files.

**Reply by valadas** [reply]
My guess is:
3rd party developers may have some dlls in common, let's say a third party library like EntityFramework that many 3rd parties could have as a dependency. DNN has an automation where if 2 extensions rely on the same library with different versions, it creates binding redirects so that 99% of the time both modules would still work but use the latest version of that library. It's not 100% but at least it covers most scenarios and should there be a breaking change, then it would only affect the older module. So I guess this is a protection to have the module using best pratices (declaring the version) win so that another module not using best practices does not break the one that does...

Long story short, do declare the version for any dll, it is best practice 😄
